### PR TITLE
Add and Enable IM subscription support

### PR DESCRIPTION
**Pull request by yunhanw-google**
#### Problem
Follow IM spec chapter 5 and IM encoding spec 7.3 and 7.4 to achieve
subscription functionality.
https://github.com/CHIP-Specifications/connectedhomeip-spec/blob/master/src/data_model/Interaction-Model.adoc#5-subscribe-interaction
https://github.com/CHIP-Specifications/connectedhomeip-spec/blob/master/src/data_model/Encoding-Specification.adoc#subscriberequestmessage
 #7978  #7979
#### Change overview
- Establishing distinct subscriptions and tracking their respective configurations from multiple clients simultaneously.
- Sharing of reporting logic between reads and subscriptions to maximize re-use.
- Multiple paths per subscription
- Automatic liveness assessment on client and server and termination of subscription accordingly.
- Eventing support in addition to attribute reporting.
- Spec compliance to the encoding specification with TLV-based payloads.
- Preservation of the application level APIs for client-side reporting, but with a swap of the internals to use the new machinery
- Update templates to use interaction model APIs for sending subscribe
requests.
- Update DeviceControllerInteractionModelDelegate to handle report
messages from subscription.
- Update ChipCallback Mgr to bridge subscribe responses to existing
callbacks, and handle TLV message just as read responses.

#### Testing
- Update TestSuite for basic Subscription protocol tests, test routing
looks like follows:
Send subscribe request, wait for response.
Wait for a few seconds, and execute "kick" commands.
The test is expected to receive the same number of Reports as the
  number of kick commands.
The test will fail when not receiving enough report data from server
  in global timeout.
- Update python script test to (in cirque) include a similar test
routine as test suite.
- Add unit test for positive and negative subscribe test
- Add integrated cirque test for subscribe

### CODE REVIEW

**Comment by andy31415** [comment]
Marked discussion required - this is one of the proposals for subscription support (based on read extension to support subscriptions)

**Comment by woody-apple** [comment]
Per discussion today this is now SDK Approved, and we should proceed with this path forward. Closing the alternative proposal here: https://github.com/project-chip/connectedhomeip/pull/9466

File: src/app/InteractionModelEngine.cpp [hunk 1]
```diff
@@ -289,6 +293,42 @@ CHIP_ERROR InteractionModelEngine::OnWriteRequest(Messaging::ExchangeContext * a
...
+        if (!readClient.IsMatchingClient(subscriptionId))
+        {
+            continue;
+        }
+
+        readClient.SetExchangeContext(apExchangeContext);
+        CHIP_ERROR err = readClient.ProcessReportData(std::move(aPayload));
+        if (err != CHIP_NO_ERROR)
+        {
+            readClient.Shutdown();
```
Comment (andy31415) [inline src/app/InteractionModelEngine.cpp:325]: Is there a situation where we want 'error in processreportdata should not shutdown'?

I would find a `return readClient.ProcessReportData(std::move(aPayload));` more obvious to read than this.

File: src/app/InteractionModelEngine.cpp [hunk 2]
```diff
@@ -329,6 +379,19 @@ CHIP_ERROR InteractionModelEngine::SendReadRequest(ReadPrepareParams & aReadPrep
     return err;
 }
 
+CHIP_ERROR InteractionModelEngine::SendSubscribeRequest(ReadPrepareParams & aReadPrepareParams, uint64_t aAppIdentifier)
+{
+    ReadClient * client = nullptr;
+    CHIP_ERROR err      = CHIP_NO_ERROR;
+    ReturnErrorOnFailure(NewReadClient(&client, ReadClient::InteractionType::Subscribe, aAppIdentifier));
+    err = client->SendSubscribeRequest(aReadPrepareParams);
```
Comment (andy31415) [inline src/app/InteractionModelEngine.cpp:387]: similar here: a pattern of 'on error the caller is responsible for shutdown' seems not as convenient as 'call the method and be done'.

I can see spearating shutdown lifetime from other operations as providing more control, but wondering if it is useful. A blanket 'on any error, client becomes invalid' seems to give some consistency and maybe simplify APIs.

This can be done in future PRs though. I would prefer not to completely derail something that works.

File: src/app/ReadClient.cpp [hunk 3]
```diff
@@ -459,5 +523,141 @@ CHIP_ERROR ReadClient::ProcessAttributeDataList(TLV::TLVReader & aAttributeDataL
...
+    VerifyOrExit(mpExchangeCtx != nullptr, err = CHIP_ERROR_NO_MEMORY);
+    mpExchangeCtx->SetResponseTimeout(kImMessageTimeoutMsec);
+
+    err = mpExchangeCtx->SendMessage(Protocols::InteractionModel::MsgType::SubscribeRequest, std::move(msgBuf),
+                                     Messaging::SendFlags(Messaging::SendMessageFlags::kExpectResponse));
+    SuccessOrExit(err);
+    MoveToState(ClientState::AwaitingInitialReport);
+
+exit:
+    ChipLogFunctError(err);
```
Comment (andy31415) [inline src/app/ReadClient.cpp:652]: I wonder if we could remove all ChipLogFunctError calls in this PR. A separate PR #9512 tries to remove them in the codebase already.

**Review by andy31415 — APPROVED** [review]

File: src/app/ReadClient.cpp (hunk 3 above)
Comment (yunhanw-google) [reply src/app/ReadClient.cpp:652]: removed, done, thanks

File: src/app/InteractionModelEngine.cpp (hunk 2 above)
Comment (yunhanw-google) [reply src/app/InteractionModelEngine.cpp:387]: move shutdown inside SendSubscribeRequest, thanks

File: src/app/InteractionModelEngine.cpp (hunk 1 above)
Comment (yunhanw-google) [reply src/app/InteractionModelEngine.cpp:325]: move shutdown in ProcessReportData , the lifetime is controlled inside readclient, done, thanks

File: src/app/reporting/Engine.cpp [hunk 4]
```diff
@@ -336,6 +367,66 @@ void Engine::Run()
...
+    }
+    if (!InteractionModelEngine::GetInstance()->MergeOverlappedAttributePath(mpChangePathList, aClusterInfo) &&
+        InteractionModelEngine::GetInstance()->IsOverlappedAttributePath(aClusterInfo))
+    {
+        ReturnLogErrorOnFailure(InteractionModelEngine::GetInstance()->PushFront(mpChangePathList, aClusterInfo));
+    }
+    return CHIP_NO_ERROR;
+}
+
+void Engine::UpdateReadHandlerDirty(ReadHandler & aReadHandler)
```
Comment (mrjerryjohns) [inline src/app/reporting/Engine.cpp:405]: Please add a comment about what this is doing in the header file.

File: src/app/reporting/Engine.cpp [hunk 5]
```diff
@@ -336,6 +367,66 @@ void Engine::Run()
         mCurReadHandlerIdx = (mCurReadHandlerIdx + 1) % CHIP_IM_MAX_NUM_READ_HANDLER;
         readHandler        = imEngine->mReadHandlers + mCurReadHandlerIdx;
     }
+
+    bool allReadClean = true;
+    for (auto & handler : InteractionModelEngine::GetInstance()->mReadHandlers)
```
Comment (mrjerryjohns) [inline src/app/reporting/Engine.cpp:372]: I think there is a case here, where on every iteration through `Run()`, you're might end up in a situation where you'll keep generating the same report to a particular subscriber, and never ever progress to generating reports for other subscribers.

The situation happens I think, when:
* You were not able to finish dispatching all reports to all subscribers in a single execution of `Run()`. Consequently, the global change set is unchanged upon completion of `Run()`.
* You were not able to finish dispatching reports because you were not able to fit all attribute data into a single message. `BuildSingleReportDataAttributeDataList` errors out, but leaves its dirty bit enabled on the ReadHandler.

On the next run, it's going to go through the same motions, error'ing out building a report to the same handler.

Don't think this is a use-case that will be encountered immediately, but it's worth thinking filing it as a follow-up issue/PR.

File: src/app/reporting/Engine.h [hunk 6]
```diff
@@ -119,6 +127,8 @@ class Engine
      *
      */
     uint32_t mCurReadHandlerIdx = 0;
+
+    ClusterInfo * mpChangePathList = nullptr;
```
Comment (mrjerryjohns) [inline src/app/reporting/Engine.h:131]:
```suggestion
    ClusterInfo * mpGlobalChangeList = nullptr;
```

You have a mixture of methods that have the term 'dirty' in them, and a mixture that use 'Change' as the term. Please pick one or the other (my vote: dirty, since it has a basis in caches and is well understood).

Comment (mrjerryjohns) [inline src/app/reporting/Engine.h:131]: I also don't think it's a wise idea to have this global dirty rely on the same ClusterInfo resource pool that is used for allocating paths on incoming subscribe/read processing. This couples one onto the other - if there are a large number of subscribers who use up the majority of the pool, your dirty tracking system is not going to function well, if it all.

They have separate purposes, so having separate pools would be more sensible. Fine to do this as a follow-up issue.

Comment (mrjerryjohns) [inline src/app/reporting/Engine.h:131]: `ClusterInfo` unfortunately is not a very good name for this type. It really should be an `AttributePath` (to match the terminology in the spec).

File: src/app/reporting/Engine.h [hunk 7]
```diff
@@ -77,6 +79,11 @@ class Engine
      */
     CHIP_ERROR ScheduleRun();
 
+    /**
+     * Application marks mutated change path and would be sent out in later report.
+     */
+    CHIP_ERROR SetDirty(ClusterInfo & aClusterInfo);
```
Comment (mrjerryjohns) [inline src/app/reporting/Engine.h:85]: Inconsistent naming with `mpChangePathList`.

File: src/app/reporting/Engine.h (hunk 6 above)
Comment (mrjerryjohns) [inline src/app/reporting/Engine.h:131]: Need comments on what this variable is tracking, and why it exists. This is a very key, important list, and the distinction between this and the ClusterInfo list tracked in each ReadHandler has to be called out.

Comment (yunhanw-google) [reply src/app/reporting/Engine.h:131]: sure, would follow-up it to create the global dirty path list with object pool, https://github.com/project-chip/connectedhomeip/issues/9551

Comment (yunhanw-google) [reply src/app/reporting/Engine.h:131]: Just update it with mpGlobalDirtyList

Comment (yunhanw-google) [reply src/app/reporting/Engine.h:131]: I would follow-up it in https://github.com/project-chip/connectedhomeip/issues/9552

File: src/app/reporting/Engine.h (hunk 7 above)
Comment (yunhanw-google) [reply src/app/reporting/Engine.h:85]: Yep, update it with mpGlobalDirtyList

File: src/app/reporting/Engine.cpp (hunk 4 above)
Comment (yunhanw-google) [reply src/app/reporting/Engine.cpp:405]: Done, thanks

File: src/app/reporting/Engine.h (hunk 6 above)
Comment (yunhanw-google) [reply src/app/reporting/Engine.h:131]: Just added, thanks

File: src/app/reporting/Engine.cpp (hunk 5 above)
Comment (yunhanw-google) [reply src/app/reporting/Engine.cpp:372]: Yep, file the ticket here, https://github.com/project-chip/connectedhomeip/issues/9553, thanks

**Review by jmartinez-silabs — APPROVED** [review]

File: src/app/ReadClient.cpp [hunk 8]
```diff
@@ -459,5 +535,141 @@ CHIP_ERROR ReadClient::ProcessAttributeDataList(TLV::TLVReader & aAttributeDataL
...
+    ReadClient * const client = reinterpret_cast<ReadClient *>(apAppState);
+    ChipLogError(DataManagement, "Subscription Liveness timeout, shutting down");
+    if (client->IsFree())
+    {
+        ChipLogError(DataManagement,
+                     "ReadClient::OnLivenessTimeoutCallback invoked on a free client! This is a bug in CHIP stack!");
+        return;
+    }
+    // TODO: add a more specific error here for liveness timeout failure to distinguish between other classes of timeouts (i.e
+    // response timeouts).
```
Comment (msandstedt) [inline src/app/ReadClient.cpp:570]: Are we saying we intend to distinguish between 'did not get ack for StatusReport' and 'did not get new ReportDataMessage or sync'?

If so, I think we've gone a little over the top with all of these transactions flying back and forth.  Why don't we just send ReportDataMessage with MRP.  If no ack is received, the sender knows its peer is gone.  What is the point of the StatusReport?

Comment (yunhanw-google) [reply src/app/ReadClient.cpp:570]: status report is used to send reply message from client to server when client receive report message, and can convey IM status code info to tell server what happend in IM layer. In addition MRP cannot be used in BLE.

**Review by hubTab — APPROVED** [review]
Code reviewed.

**Review by mrjerryjohns — APPROVED** [review]

File: examples/chip-tool/templates/reporting-commands.zapt [hunk 9]
```diff
@@ -31,7 +31,7 @@ public:
 {{#chip_client_clusters}}
 {{#chip_server_cluster_attributes}}
 {{#if isReportableAttribute}}
-        callbacksMgr.AddReportCallback(GetExecContext()->storage->GetRemoteNodeId(), endpointId, {{asHex parent.code 4}}, {{asHex code 4}}, onReport{{asUpperCamelCase parent.name}}{{asUpperCamelCase name}}Callback->Cancel());
+        callbacksMgr.AddReportCallback(GetExecContext()->storage->GetRemoteNodeId(), endpointId, {{asHex parent.code 4}}, {{asHex code 4}}, onReport{{asUpperCamelCase parent.name}}{{asUpperCamelCase name}}Callback->Cancel(), BasicAttributeFilter<{{chipCallback.name}}AttributeCallback>);
```
Comment (vivien-apple) [inline examples/chip-tool/templates/reporting-commands.zapt:34]: You may want to check if the attribute is a `List` to set up the right filter.

File: src/app/zap-templates/templates/app/CHIPClusters-src.zapt [hunk 10]
```diff
@@ -161,24 +159,18 @@ CHIP_ERROR {{asUpperCamelCase parent.name}}Cluster::WriteAttribute{{asUpperCamel
...
-       .Put16(maxInterval);
-    {{#if isAnalog}}
-    buf.Put{{chipTypePutLength}}(static_cast<{{chipTypePutCastType}}>(change));
-    {{/if}}
-    COMMAND_FOOTER();
+    {{#if isAnalog}}IgnoreUnusedVariable(change);{{/if}}
+    chip::app::AttributePathParams attributePath;
+    attributePath.mEndpointId = mEndpoint;
+    attributePath.mClusterId  = mClusterId;
+    attributePath.mFieldId    = {{asHex code 8}};
```
Comment (vivien-apple) [inline src/app/zap-templates/templates/app/CHIPClusters-src.zapt:166]: nit: `{{asHex code 8}}` -> `{{#if isGlobalAttribute}}Globals{{else}}{{asUpperCamelCase parent.name}}{{/if}}::Attributes::Ids::{{asUpperCamelCase name}}`

File: src/app/zap-templates/templates/app/CHIPClusters-src.zapt [hunk 11]
```diff
@@ -161,24 +159,18 @@ CHIP_ERROR {{asUpperCamelCase parent.name}}Cluster::WriteAttribute{{asUpperCamel
...
+    attributePath.mFlags.Set(chip::app::AttributePathParams::Flags::kFieldIdValid);
+    return mDevice->SendSubscribeAttributeRequest(attributePath, minInterval, maxInterval, onSuccessCallback, onFailureCallback);
 }
 
 CHIP_ERROR {{asUpperCamelCase parent.name}}Cluster::ReportAttribute{{asUpperCamelCase name}}(Callback::Cancelable * onReportCallback)
 {
-    return RequestAttributeReporting({{asHex code 4}}, onReportCallback);
+    return RequestAttributeReporting({{asHex code 4}}, onReportCallback, BasicAttributeFilter<{{chipCallback.name}}AttributeCallback>);
```
Comment (vivien-apple) [inline src/app/zap-templates/templates/app/CHIPClusters-src.zapt:173]: `{{asHex code 4}}` -> {{#if isGlobalAttribute}}Globals{{else}}{{asUpperCamelCase parent.name}}{{/if}}::Attributes::Ids::{{asUpperCamelCase name}}`

File: src/controller/CHIPDevice.cpp [hunk 12]
```diff
@@ -720,6 +720,45 @@ CHIP_ERROR Device::SendReadAttributeRequest(app::AttributePathParams aPath, Call
...
+
+    ReturnErrorOnFailure(LoadSecureSessionParametersIfNeeded(loadedSecureSession));
+
+    app::AttributePathParams * path = mpIMDelegate->AllocateAttributePathParam(1, seqNum);
+
+    VerifyOrReturnError(path != nullptr, CHIP_ERROR_NO_MEMORY);
+
+    *path = aPath;
+
+    // The application context is used to identify different requests from client applicaiton the type of it is intptr_t, here we
```
Comment (vivien-apple) [inline src/controller/CHIPDevice.cpp:739]: nit: applicaiton

**Review by vivien-apple — COMMENTED** [review]
As mentioned previously, I would have preferred this PR to be broken down into smaller steps. It is hard for me to understand properly all the states machines, and the ramifications of the changes.

My main concern would be the yaml test 
 1. how it works in order to ensure it will be easy to write test cases anytime we discover an issue. I know @erjiaqing is working on writing it differently
 2. Add back an updated Darwin test

File: src/app/ReadClient.h [hunk 13]
```diff
@@ -84,11 +103,17 @@ class ReadClient : public Messaging::ExchangeDelegate
...
     enum class ClientState
     {
-        Uninitialized = 0,     ///< The client has not been initialized
-        Initialized,           ///< The client has been initialized and is ready for a SendReadRequest
-        AwaitingInitialReport, ///< The client is waiting for initial report
+        Uninitialized = 0,         ///< The client has not been initialized
+        Initialized,               ///< The client has been initialized and is ready for a SendReadRequest
+        AwaitingInitialReport,     ///< The client is waiting for initial report
+        AwaitingSubscribeResponse, ///< The client is waiting for subscribe response
+        SubscriptionIdle,          ///< The client is maintaining subscription
```
Comment (msandstedt) [inline src/app/ReadClient.h:110]: Unless Idle comes from the spec, I suggest perhaps renaming this.  Idle to me implies we aren't doing anything.  But that's not true.  This is the active state for the subscription that we arrive at after the successful handshake.  In fact, `SubscriptionActive` might be a better name.

File: src/app/ReadClient.cpp [hunk 14]
```diff
@@ -453,5 +529,141 @@ CHIP_ERROR ReadClient::ProcessAttributeDataList(TLV::TLVReader & aAttributeDataL
...
+    ReturnLogErrorOnFailure(subscribeResponse.GetSubscriptionId(&subscriptionId));
+    VerifyOrReturnLogError(IsMatchingClient(subscriptionId), CHIP_ERROR_INVALID_ARGUMENT);
+    ReturnLogErrorOnFailure(subscribeResponse.GetMinIntervalFloorSeconds(&mMinIntervalFloorSeconds));
+    ReturnLogErrorOnFailure(subscribeResponse.GetMaxIntervalCeilingSeconds(&mMaxIntervalCeilingSeconds));
+    mpDelegate->SubscribeResponseProcessed(this);
+
+    if (IsAwaitingSubscribeResponse())
+    {
+        MoveToState(ClientState::SubscriptionIdle);
+        mpExchangeCtx = nullptr;
```
Comment (msandstedt) [inline src/app/ReadClient.cpp:591]: This is obviously the happy path at the end of the subscription setup handshake where the exchange is complete and all future messages will be unsolicited.

And apparently this is how we relinquish our exchange context and (I presume) it becomes available again for other things.

How do I evaluate the other corner case possibilities of this exchange though?  Can't the responder fail to send the expected messages or fail to send them in the expected order?  Does the code account for all of these possible cases?

File: src/app/ReadClient.cpp [hunk 15]
```diff
@@ -453,5 +529,141 @@ CHIP_ERROR ReadClient::ProcessAttributeDataList(TLV::TLVReader & aAttributeDataL
...
+    CHIP_ERROR err = CHIP_NO_ERROR;
+    System::PacketBufferHandle msgBuf;
+    System::PacketBufferTLVWriter writer;
+    SubscribeRequest::Builder request;
+    VerifyOrExit(ClientState::Initialized == mState, err = CHIP_ERROR_INCORRECT_STATE);
+    VerifyOrExit(mpDelegate != nullptr, err = CHIP_ERROR_INCORRECT_STATE);
+    msgBuf = System::PacketBufferHandle::New(kMaxSecureSduLengthBytes);
+    VerifyOrExit(!msgBuf.IsNull(), err = CHIP_ERROR_NO_MEMORY);
+
+    AbortExistingExchangeContext();
```
Comment (msandstedt) [inline src/app/ReadClient.cpp:608]: What are the implications of this automatic lifecycle management for the ReadClient?  If I was using this object, I would expect to just construct whatever number I need, and then would presume that since each client is bound to a single exchange, I would not be able to send a new request while another request is ongoing.

That is, it seems like an error to arrive here if there is an ongoing exchange.  Why are we silently aborting it?  Does this have some desirable outward facing utility or behavior?

Also, what about the case where I have an established subscription and am now waiting for unsolicited responses?  Does this still do cleanup for that case?  Will failure callbacks fire if this is called with an already established subscription.

It seems like it might just be simpler to disallow reuse of these objects while their already busy doing something.

File: src/lib/core/CHIPConfig.h [hunk 16]
```diff
@@ -2449,6 +2449,15 @@ extern const char CHIP_NON_PRODUCTION_MARKER[];
 #define CHIP_IM_MAX_NUM_WRITE_CLIENT 4
 #endif
 
+/**
+ * @def CHIP_DEVICE_CONTROLLER_SUBSCRIPTION_ATTRIBUTE_PATH_POOL_SIZE
+ *
+ * @brief Defines the object pool for allocating attribute path for subscription in device controller.
+ */
+#ifndef CHIP_DEVICE_CONTROLLER_SUBSCRIPTION_ATTRIBUTE_PATH_POOL_SIZE
+#define CHIP_DEVICE_CONTROLLER_SUBSCRIPTION_ATTRIBUTE_PATH_POOL_SIZE CHIP_IM_MAX_NUM_READ_CLIENT
```
Comment (msandstedt) [inline src/lib/core/CHIPConfig.h:2458]: For some applications, this is going to feel really restrictive.

Most of the metadata for a subscription is pretty lightweight.  What in the ReadClient is so large that we've decided we can only have 4 at a given time?  Also, is there a plan to provide a dynamic pool in applications where that makes sense?

File: src/controller/DeviceControllerInteractionModelDelegate.h [hunk 17]
```diff
@@ -0,0 +1,86 @@
...
+                return &item.Params;
+            }
+        }
+        return nullptr;
+    }
+
+private:
+    struct AttributePathTransactionMap
+    {
+        uint64_t ApplicationId = UINT64_MAX;
```
Comment (msandstedt) [inline src/controller/DeviceControllerInteractionModelDelegate.h:79]: Why do we need a map keyed on magic numerical Application IDs?  This seems like it will just force extra bookkeeping in the app.  Each subscription is a distinct instance.  Why can't callbacks just be passed a void * context?

Comment (msandstedt) [reply src/controller/DeviceControllerInteractionModelDelegate.h:79]: Maybe I'm misunderstanding.  Is this literally the SubscriptionId from the spec?  If so, that's fine I suppose.  And I presume then that we are using uint64 instead of int32 to accommodate the magic UINT64_MAX ID.

If that is so, I suggest a couple things: (1) rename this to SubscriptionId if that's what it is; (2) add a constexpr for the UINT64_MAX special value so it's clear what that means; (3) make external APIs use uint32 to prevent nonsensical Subscription IDs from being passed in or out of this class.

Comment (msandstedt) [reply src/controller/DeviceControllerInteractionModelDelegate.h:79]: Also, though not changed in this PR, app/MessageDef/ReportData.h has SubscriptionId as a uint64.  I thought for sure subscription IDs were 32bit.  Am I missing something?

File: src/app/ReadHandler.cpp [hunk 18]
```diff
@@ -128,14 +170,30 @@ CHIP_ERROR ReadHandler::OnStatusReport(Messaging::ExchangeContext * apExchangeCo
...
+    {
+        mpExchangeCtx = mpExchangeMgr->NewContext(mSessionHandle.Value(), this);
+        mpExchangeCtx->SetResponseTimeout(kImMessageTimeoutMsec);
     }
     VerifyOrReturnLogError(mpExchangeCtx != nullptr, CHIP_ERROR_INCORRECT_STATE);
-    MoveToState(HandlerState::Reporting);
-    return mpExchangeCtx->SendMessage(Protocols::InteractionModel::MsgType::ReportData, std::move(aPayload),
-                                      Messaging::SendFlags(Messaging::SendMessageFlags::kExpectResponse));
+    MoveToState(HandlerState::AwaitingReportResponse);
+    CHIP_ERROR err = mpExchangeCtx->SendMessage(Protocols::InteractionModel::MsgType::ReportData, std::move(aPayload),
```
Comment (msandstedt) [inline src/app/ReadHandler.cpp:186]: As discussed offline, when the sdk becomes spec compliant, it will be possible for the session to disappear out from under this exchange context in two scenarios: 1) session eviction and 2) session closing initiated by the peer.

This call chain will presumably fail in these cases, so we need to handle this if it is not handled already.  In this case, the subscription should be canceled.

**Comment by msandstedt** [comment]
Noted in the newly created https://github.com/project-chip/connectedhomeip/issues/9596, but worth of mention here.

We have a problem where we've organized a series of singletons, session manager -> exchange manager -> IM engine, where lower layers do not have visibility into what of their resources upper layers are reliant upon.

So for instance, if a session closes and this is not app-initiated, there is no way for the session manager to trigger a failure callback chain to release resources in the exchange layer, IM layer and app.  This can happen in at least 3 cases: (1) session eviction, (2) peer-initiated session closing, (3) session-layer-initiated closing e.g. for MRP timeout.

The subscription implementation attempts to mitigate this with the liveliness mechanism.  That is necessary and good, but it's not enough.  The problem is that this is reliant upon the report or sync cadence to release resources.  That may be too late.  For instance, imagine this sequence:


* peer initiates session
* peer initiates subscription
* [session table becomes full]
* peer closes session
* peer initiates session
* peer initiates subscription
* [oops, session table is still full]

It is not good or necessary for us to break in this case.  But to solve this, we would need the session manager to understand the resources bound to sessions and be able to clean them up by trigger a failure callback chain all the way into the app.

From a clean slate, the way to do this would be to not have exchange manager or IM singletons, but instead to have each session hold a table of exchanges and each exchange to point to IM->app context.  This would cleanly reflect the 1:many relationship between sessions and exchanges and allow for a very straight forward session-layer-initiated cleanup mechanism.  Note also that such a chain can happen cleanly even from the transport layer.  For instance, TCP connections close.  It ought to be easy to have a chain that looks like TCP close -> session close -> exchange(s) close -> IM error(s) -> app error(s).

Instead, we have an exchange manager singleton that receives the full multiplexed traffic stream from the session layer.  This creates the problem I just laid out and has the further consequence that the exchange manager must be session and peer aware.  If it was only ever possible for an exchange to be called from an associated session context, we would have a cleaner separation of the layers in our communication stack.

However, given that we do now have this exchange manager singleton and it is likely this will not change, it seems that we need some other approach.  I would recommend that we give the session manager a handle to the exchange manager and allow the session manager to notify the exchange manager when a session is evicted or disappears.  This, I hope, would give us the immediate failure callback chain we need.

**Review by msandstedt — COMMENTED** [review]
I noted some comments inline.  I also had a comment that has resulted in ticket #9596.  But on the whole, I think this is really good code.  It's a big feature, and this seems to address things pretty cleanly.

File: src/app/ReadClient.cpp (hunk 14 above)
Comment (yunhanw-google) [reply src/app/ReadClient.cpp:591]: yep, SubscribeResponse is last message for subscription establishment procedure, after subscription becomes active, we would expect to receive report using new exchange, so current exchange should be set nullptr. if responder fails to send the report during post-subscription, the liveness timer would fired and shutdown subscription

Comment (msandstedt) [reply src/app/ReadClient.cpp:591]: Yes, but what happens if the SubscribeResponse is never received?  And what happens if SubscribeResponse is received before the initial report?  Seeing only one place where the exchange is closed leads me to be concerned that we don't have logic to handle other error cases where it should also be closed.

File: src/app/ReadClient.cpp (hunk 15 above)
Comment (yunhanw-google) [reply src/app/ReadClient.cpp:608]: I think it is fine to remove this line. before sending subscribe request, we expect exchange has been cleared, if not, it is buggy.

File: src/controller/DeviceControllerInteractionModelDelegate.h (hunk 17 above)
Comment (yunhanw-google) [reply src/controller/DeviceControllerInteractionModelDelegate.h:79]: in latest spec, refreshed on this monday, this becomes 32bit, I would follow-up it immediately after this pr is landed on this ticket, https://github.com/project-chip/connectedhomeip/issues/9520

File: src/app/ReadHandler.cpp (hunk 18 above)
Comment (yunhanw-google) [reply src/app/ReadHandler.cpp:186]: filed #9596,

File: src/lib/core/CHIPConfig.h (hunk 16 above)
Comment (yunhanw-google) [reply src/lib/core/CHIPConfig.h:2458]: yes, agree. we would use object pool for several pool in interaction model engine in follow-up next week. for this one, we would also use object pool as well, file ticket https://github.com/project-chip/connectedhomeip/issues/9602

File: src/app/ReadClient.h (hunk 13 above)
Comment (yunhanw-google) [reply src/app/ReadClient.h]: Done, thanks

File: src/controller/CHIPDevice.cpp (hunk 12 above)
Comment (yunhanw-google) [reply src/controller/CHIPDevice.cpp]: Done, thanks

File: src/app/ReadClient.cpp (hunk 14 above)
Comment (yunhanw-google) [reply src/app/ReadClient.cpp]: yep, agree, thanks for catching up this corner case, 
add IsAwaitingSubscribeResponse() check before process ProcessSubscribeResponse to handle the situation for SubscribeResponse is received before the initial report. In addition, we don't need mark this mpExchangeCtx = nullptr; since when message timeout, shutdown is calle,d we would clear mpExchangeCtx in shutdown

**Review by msandstedt — APPROVED** [review]

File: src/app/zap-templates/templates/app/CHIPClusters-src.zapt (hunk 11 above)
Comment (yunhanw-google) [reply src/app/zap-templates/templates/app/CHIPClusters-src.zapt:173]: Done, thanks

File: src/app/zap-templates/templates/app/CHIPClusters-src.zapt (hunk 10 above)
Comment (yunhanw-google) [reply src/app/zap-templates/templates/app/CHIPClusters-src.zapt:166]: Done, thanks

File: src/app/ReadClient.h (hunk 13 above)
Comment (erjiaqing) [reply src/app/ReadClient.h]: How about `SubscriptionAlive`, since `Active` maybe intepreted as "we are sending reports now"

File: src/lib/core/CHIPConfig.h (hunk 16 above)
Comment (erjiaqing) [reply src/lib/core/CHIPConfig.h:2458]: As this lives in `CHIPConfig.h`, this is designed to be a default value.
There will be follow ups for dynamic pool in applications (https://github.com/project-chip/connectedhomeip/issues/9602).

File: src/controller/DeviceControllerInteractionModelDelegate.h (hunk 17 above)
Comment (erjiaqing) [reply src/controller/DeviceControllerInteractionModelDelegate.h:79]: The case is we don't want to expose the underlying objects to the user for memory safety. (i.e. the other commands will use a handle wrapper when possible, and for read / subscriptions, as the input is simplier, we can call the im wrapper directly without using handle)

Also the pointer can be unreliable for classes  inheritance relations (since a pointer pointing to base class and child class  can have different values),.

As all callbacks in interaction model delegate has a pointer to the objects, users CAN use them directly without using the appid.

File: examples/chip-tool/templates/reporting-commands.zapt (hunk 9 above)
Comment (yunhanw-google) [reply examples/chip-tool/templates/reporting-commands.zapt:34]: https://github.com/project-chip/connectedhomeip/issues/9643, thanks song would follow-up it in this ticket, thanks

**Comment by yunhanw-google** [comment]
@vivien-apple @erjiaqing has updated the test, any further concern, thanks

**Comment by vivien-apple** [comment]
Here is how I think we should write the test:

```
name: Subscribe Tests - OnOff Cluster

 config:
     cluster: "On/Off"
     endpoint: 1

 tests:
     - label: "Set OnOff Attribute to false"
       command: "Off"

     - label: "Subscribe OnOff Attribute"
       command: "subscribeAttribute"
       attribute: "OnOff"
       minInterval: 2
       maxInterval: 10
       response:
            value: false

     - label: "Set OnOff Attribute to true"
       command: "On"

     - label: "Wait for OnOff attribute report: True"
       command: "waitForAttributeReport"
       attribute: "OnOff"
       response:
           value: true

     - label: "Set OnOff Attribute to false"
       command: "Off"

     - label: "Wait for OnOff attribute report: False"
       command: "waitForAttributeReport"
       attribute: "OnOff"
       response:
           value: false
```

I have changed a little bit the syntax for "waitForAttributeReport" as it can probably be done without having to resort to touching the "simulated" cluster bits and that would let the command inherited `cluster` and `endpoint`

File: src/app/ReadClient.h [hunk 19]
```diff
@@ -115,16 +141,21 @@ class ReadClient : public Messaging::ExchangeDelegate
      *
      */
     bool IsFree() const { return mState == ClientState::Uninitialized; }
+    bool IsSubscriptionTypeIdle() const { return mState == ClientState::SubscriptionActive; }
     bool IsAwaitingInitialReport() const { return mState == ClientState::AwaitingInitialReport; }
+    bool IsAwaitingSubscribeResponse() const { return mState == ClientState::AwaitingSubscribeResponse; }
+
     CHIP_ERROR GenerateEventPathList(EventPathList::Builder & aEventPathListBuilder, EventPathParams * apEventPathParamsList,
                                      size_t aEventPathParamsListSize);
     CHIP_ERROR GenerateAttributePathList(AttributePathList::Builder & aAttributeathListBuilder,
```
Comment (hubTab) [inline src/app/ReadClient.h:150]: SmartThings is very excited to start integrating Subscription Interactions, and this PR is an excellent path forward. 
In fact, we have started already using the interactions, by using the Read Interaction, which is based on lib/connectedhomeip/src/app/tests. However, we run into a couple of issues with the ReadClient class:

Currently ReadClient.h has GenerateAttributePathList(), ProcessReportData() and the destructor (virtual ~ReadClient()) marked as private, making them inaccessible for use outside the class. To get past this hurdle, we have declared them as public so that our application can use the ReadClient class.

I suggest setting them up as public, so that the SDK consumer can use the ReadClient class, in a similar way the TestReadInteraction (as it is set as a friend class, but this would not be a path moving forward for custom code). Or if this is not preferable, we are open to alternatives that you can suggest so Read Interaction development can be continue.
